v2/component/launcher.go

```diff
@@ -178,100 +182,125 @@ func NewLauncher(runtimeInfo string, options *LauncherOptions) (*Launcher, error
+	for name, artifactList := range executorInput.Inputs.Artifacts {
+		// TODO(neuromage): Support concat-based placholders for arguments.
+		if len(artifactList.Artifacts) == 0 {
+			continue
 		}
+		inputArtifact := artifactList.Artifacts[0]
 
-		b, err := ioutil.ReadFile(v.FileInputPath)
-		if err != nil {
-			return fmt.Errorf("Failed to read input artifact metadata file for %q: %v", k, err)
-		}
+		// Prepare input uri placeholder.
+		key := fmt.Sprintf(`{{$.inputs.artifacts['%s'].uri}}`, name)
+		l.placeholderReplacements[key] = inputArtifact.Uri
 
-		a := &pb.Artifact{}
-		if err := protojson.Unmarshal(b, a); err != nil {
-			return fmt.Errorf("Failed to unmarshall input artifact metadata for %q: %v", k, err)
+		localPath, err := localPathForURI(inputArtifact.Uri)
+		if err != nil {
+			glog.Warningf("Input Artifact %q does not have a recognized storage URI %q. Skipping downloading to local path.", name, inputArtifact.Uri)
+			continue
 		}
 
-		v.Artifact = a
-
-		// TODO(neuromage): Support `{{$}}` placeholder for components using ExecutorInput.
-		// TODO(neuromage): Support concat-based placholders for arguments.
-
-		// Prepare input uri placeholder.
-		key := fmt.Sprintf(`{{$.inputs.artifacts['%s'].uri}}`, k)
-		l.placeholderReplacements[key] = v.Artifact.GetUri()
-
 		// Prepare input path placeholder.
-		v.LocalArtifactFilePath = path.Join("/tmp/kfp_launcher_inputs", k, "data")
-		key = fmt.Sprintf(`{{$.inputs.artifacts['%s'].path}}`, k)
-		l.placeholderReplacements[key] = v.LocalArtifactFilePath
+		key = fmt.Sprintf(`{{$.inputs.artifacts['%s'].path}}`, name)
+		l.placeholderReplacements[key] = localPath
 
 		// Copy artifact to local storage.
+		copyErr := func(err error) error {
+			return fmt.Errorf("failed to download input artifact %q from remote storage URI %q: %w", name, inputArtifact.Uri, err)
+		}
 		// TODO: Selectively copy artifacts for which .path was actually specified
 		// on the command line.
-
-		blobKey, err := l.bucketConfig.keyFromURI(v.Artifact.GetUri())
+		blobKey, err := l.bucketConfig.keyFromURI(inputArtifact.Uri)
 		if err != nil {
-			return err
+			return copyErr(err)
 		}
-		r, err := bucket.NewReader(ctx, blobKey, nil)
-		if err != nil {
-			return err
-		}
-		defer r.Close()
 
-		if err := os.MkdirAll(path.Dir(v.LocalArtifactFilePath), 0644); err != nil {
-			return err
-		}
-		w, err := os.Create(v.LocalArtifactFilePath)
-		if err != nil {
-			return err
-		}
-		if _, err := io.Copy(w, r); err != nil {
-			return err
+		if err := downloadBlob(ctx, bucket, localPath, blobKey); err != nil {
+			return copyErr(err)
 		}
 	}
 
 	// Prepare input parameter placeholders.
-	for k, v := range l.runtimeInfo.InputParameters {
-		key := fmt.Sprintf(`{{$.inputs.parameters['%s']}}`, k)
-		l.placeholderReplacements[key] = v.ParameterValue
+	for name, parameter := range executorInput.Inputs.Parameters {
+		key := fmt.Sprintf(`{{$.inputs.parameters['%s']}}`, name)
```

Comment:
https://4e18c21c9d33d20f-dot-datalab-vm-staging.googleusercontent.com/#/runs/details/083e11c2-db45-462c-ab4b-2321149c21b7

In this placeholder concat sample, I'm still seeing the placeholder not replaced.
Can you verify its sample and test (we might want to change verify function to confirm the output value)?
https://github.com/kubeflow/pipelines/blob/c9b20692f316b325098e74a9dc6caa5296a2d0dd/samples/test/placeholder_concat.py

Reply:
Done.